code/obj/submachine/cooking.dm

```diff
@@ -1573,53 +1576,82 @@ TYPEINFO(/obj/submachine/mixer)
 		user.u_equip(W)
 		W.set_loc(src)
 		W.dropped(user)
+		tgui_process.update_uis(src)
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "MixerMachine")
+			ui.open()
+	ui_static_data(mob/user)
+		. = list("maxItems" = MIXER_MAX_CONTENTS)
+
+	ui_data(mob/user)
+		var/mixerContents = list()
+		var/index = 1
+		for(var/obj/item/I in src.contents)
+			var/itemData = list()
+			itemData["name"] = I.name
+			itemData["index"] = index
+			itemData["iconData"] = get_item_icon(I)
+			mixerContents += list(itemData)
+			index += 1
+		. = list("working" = src.working, "mixerContents" = mixerContents)
+
+	proc/get_item_icon(var/obj/item/target)
+		var/static/base64_preview_cache = list()
+		var/original_name = initial(target.name)
+		. = base64_preview_cache[original_name]
+
+		if(isnull(.))
+			var/icon/result = getFlatIcon(target, no_anim=TRUE)
+			if(result)
+				. = icon2base64(result)
+			else
+				. = ""
+			base64_preview_cache[original_name] = .
 
-	attack_hand(var/mob/user)
-		if (!src.working)
-			src.add_dialog(user)
-			var/dat = {"<B>KitchenHelper Mixer</B><BR>
-			<HR>
-			<B>Contents:</B><BR>"}
-			for (var/obj/item/I in src.contents)
-				dat += "[I]<BR>"
-			dat += {"<HR>
-			<A href='?src=\ref[src];mix=1'>Mix!</A><BR>
-			<A href='?src=\ref[src];eject=1'>Eject Contents</A>"}
-			user.Browse(dat, "window=mixer;size=400x500")
-			onclose(user, "mixer")
-		else
-			src.add_dialog(user)
-			var/dat = {"<B>KitchenHelper Mixer</B><BR>
-			<HR><BR>
-			Mixing! Please wait!"}
-			user.Browse(dat, "window=mixer;size=400x500")
-			onclose(user, "mixer")
+	proc/ejectItemFromMixer(obj/item/target)
+		if (target)
+			if (BOUNDS_DIST(usr, src) == 0)
+				usr.put_in_hand_or_drop(target)
+			else
+				target.set_loc(src.loc)
+
+	ui_act(action, params)
+		. = ..()
+		if (.)
+			return
+
+		switch (action)
+			if ("eject")
+				var/index = params["index"]
+				var/obj/item/target = src.contents[index]
+				src.ejectItemFromMixer(target)
+
+				usr.show_text(SPAN_NOTICE("You eject the [target.name] from the [src]."))
+				. = TRUE
+
+			if ("mix")
+				src.mix()
+				. = TRUE
+
+			if ("ejectAll")
+				for (var/obj/item/target in src.contents)
+					src.ejectItemFromMixer(target)
+
+				usr.show_text(SPAN_NOTICE("You eject all contents from the [src]."))
+		if (.)
+			tgui_process.update_uis(src)
```

Comment:
```suggestion
```
Returning true from this function already updates the UI, you don't need to do it explicitly

Reply:
Removed